slang.h

```diff
@@ -2543,8 +2597,329 @@ namespace slang
+
+            /** Specialize a type based on type arguments.
+            */
+        virtual SLANG_NO_THROW TypeReflection* SLANG_MCALL specializeType(
+            TypeReflection*             type,
+            SlangInt                    specializationArgCount,
+            SpecializationArg const*    specializationArgs,
+            ISlangBlob**                outDiagnostics = nullptr) = 0;
+
+
+            /** Get the layout `type` on the chosen `target`.
+            */
+        virtual SLANG_NO_THROW TypeLayoutReflection* SLANG_MCALL getTypeLayout(
+            TypeReflection* type,
+            SlangInt        targetIndex = 0,
+            LayoutRules     rules = LayoutRules::Default,
+            ISlangBlob**    outDiagnostics = nullptr) = 0;
+
+            /** Create a request to load/compile front-end code.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL createCompileRequest(
+            SlangCompileRequest**   outCompileRequest) = 0;
+
+#if 0
+            /** Add a path to use when searching for `#include`d or `import`ed files.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL addSearchPath(
+            char const* path) = 0;
+
+            /** Add a global preprocessor definition to use for all compilation.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL addPreprocessorDefine(
+            char const* name,
+            char const* value) = 0;
+
+            /** Set the default matrix layout mode to use for all compiles.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL setMatrixLayoutMode(
+            SlangMatrixLayoutMode mode) = 0;
+#endif
+    };
+
+    #define SLANG_UUID_ISession { 0x67618701, 0xd116, 0x468f, { 0xab, 0x3b, 0x47, 0x4b, 0xed, 0xce, 0xe, 0x3d } }
+
+
+        /** A module is the granularity of shader code compilation and loading.
+
+        In most cases a module corresponds to a single compile "translation unit."
+        This will often be a single `.slang` or `.hlsl` file and everything it
+        `#include`s.
+
+        Notably, a module `M` does *not* include the things it `import`s, as these
+        as distinct modules that `M` depends on. There is a directed graph of
+        module dependencies, and all modules in the graph must belong to the
+        same session (`ISession`).
+        */
+    struct IModule : public ISlangUnknown
+    {
+    public:
+    };
+    
+    #define SLANG_UUID_IModule { 0xc720e64, 0x8722, 0x4d31, { 0x89, 0x90, 0x63, 0x8a, 0x98, 0xb1, 0xc2, 0x79 } }
+
+
+        /** Argument used for specialization to types/values.
+        */
+    struct SpecializationArg
+    {
+        enum class Kind : int32_t
+        {
+            Unknown,
+            Type,
+        };
+        Kind kind;
+        union
+        {
+            TypeReflection* type;
+        };
+    };
+
+        /** Description of a program to be created.
+        */
+    struct ProgramDesc
```

Comment:
This may be best way to do this. But if there is a base interface (perhaps having no new methods) that both inherited from, then perhaps it might be cleaner to just use an array of pointers to that type, and use dynamic casting internally.

Reply:
That's a cleanup I think will happen automatically if I do the big refactor I'm thinking of. If I introduce a base interface for (lacking a better term) "component types," then a component type would be the base interface, and all of the current "module," "entry point," and "program" concepts would fit into it. Also the new "entry point group" concept would probably not need to be a separate thing. The main catch would be that the new concept probably ends up needing its own reflection API...